x-pack/plugins/security/public/management/roles/edit_role/privileges/kibana/privilege_form_calculator/privilege_form_calculator.ts

```diff
@@ -0,0 +1,295 @@
+   */
+  public updateSelectedFeaturePrivilegesForCustomization(
+    featureId: string,
+    privilegeIndex: number,
+    willBeCustomizing: boolean
+  ) {
+    const primary = this.getDisplayedPrimaryFeaturePrivilege(featureId, privilegeIndex);
+    const selectedFeaturePrivileges = this.getSelectedFeaturePrivileges(featureId, privilegeIndex);
+
+    if (!primary) {
+      return selectedFeaturePrivileges;
+    }
+
+    const nextPrivileges = [];
+
+    if (willBeCustomizing) {
+      const feature = this.kibanaPrivileges.getSecuredFeature(featureId);
+
+      const startingPrivileges = feature
+        .getSubFeaturePrivileges()
+        .filter(ap => primary.grantsPrivilege(ap))
+        .map(p => p.id);
+
+      const existingPrivileges = selectedFeaturePrivileges.filter(
+        sfp => sfp !== primary.id && sfp !== primary.getCorrespondingPrivilegeId()
+      );
+
+      nextPrivileges.push(
+        ...existingPrivileges,
+        primary.getCorrespondingPrivilegeId(),
+        ...startingPrivileges
+      );
+    } else {
+      nextPrivileges.push(primary.id);
+    }
+
+    return nextPrivileges;
+  }
+
+  /**
+   * Determines if the indicated privilege entry is less permissive than the configured "global" entry for the role.
+   * @param privilegeIndex the index of the kibana privileges role component
+   */
+  public hasSupersededInheritedPrivileges(privilegeIndex: number) {
+    const global = this.locateGlobalPrivilege(this.role);
+
+    const entry = this.role.kibana[privilegeIndex];
+
+    if (isGlobalPrivilegeDefinition(entry) || !global) {
+      return false;
+    }
+
+    const globalPrivileges = this.kibanaPrivileges.createCollectionFromRoleKibanaPrivileges([
+      global,
+    ]);
+
+    const formPrivileges = this.kibanaPrivileges.createCollectionFromRoleKibanaPrivileges([
+      this.role.kibana[privilegeIndex],
+    ]);
+
+    const hasAssignedBasePrivileges = this.kibanaPrivileges
+      .getBasePrivileges(entry)
+      .some(base => entry.base.includes(base.id));
+
+    const featuresWithDirectlyAssignedPrivileges = this.kibanaPrivileges
+      .getSecuredFeatures()
+      .filter(feature =>
+        feature
+          .getAllPrivileges()
+          .some(privilege => entry.feature[feature.id]?.includes(privilege.id))
+      );
+
+    const hasSupersededBasePrivileges =
+      hasAssignedBasePrivileges &&
+      this.kibanaPrivileges
+        .getBasePrivileges(entry)
+        .some(
+          privilege =>
+            globalPrivileges.grantsPrivilege(privilege) &&
+            !formPrivileges.grantsPrivilege(privilege)
+        );
+
+    const hasSupersededFeaturePrivileges = featuresWithDirectlyAssignedPrivileges.some(feature =>
+      feature
+        .getAllPrivileges()
+        .some(fp => globalPrivileges.grantsPrivilege(fp) && !formPrivileges.grantsPrivilege(fp))
+    );
+
+    return hasSupersededBasePrivileges || hasSupersededFeaturePrivileges;
+  }
+
+  public getSecuredFeatures() {
```

Comment:
nit: This isn't doing any calculations, it doesn't feel like it should be part of the calculator.

Reply:
You're right :)

It's here because I was tired of passing so many props around to various components, and since the calculator already had access to these, it was convenient to just expose them. I can revisit and see if my later cleanup efforts simplifies this any